paimon-common/src/main/java/org/apache/paimon/fileindex/bitmap/BitmapFileIndexMetaV2.java

```diff
@@ -0,0 +1,415 @@
+                                }
+                            }
+                        });
+
+        out.writeInt(indexBlocks.size());
+
+        int bitmapBodyOffset = 0;
+        for (BitmapIndexBlock e : indexBlocks) {
+            // secondary entry
+            valueWriter.accept(e.key);
+            out.writeInt(e.offset);
+            bitmapBodyOffset += e.serializedBytes;
+        }
+
+        // bitmap body offset
+        out.writeInt(bitmapBodyOffset);
+
+        // bitmap index blocks
+        for (BitmapIndexBlock indexBlock : indexBlocks) {
+            out.writeInt(indexBlock.entryList.size());
+            for (Entry e : indexBlock.entryList) {
+                valueWriter.accept(e.key);
+                out.writeInt(e.offset);
+                out.writeInt(e.length);
+            }
+        }
+    }
+
+    @Override
+    public void deserialize(SeekableInputStream seekableInputStream) throws Exception {
+
+        indexBlockStart = seekableInputStream.getPos();
+
+        InputStream inputStream = seekableInputStream;
+        if (options.getBoolean(BitmapFileIndex.ENABLE_BUFFERED_INPUT, true)) {
+            inputStream = new BufferedInputStream(inputStream);
+        }
+        DataInput in = new DataInputStream(inputStream);
+        ThrowableSupplier valueReader = getValueReader(in);
+        Function<Object, Integer> measure = getSerializeSizeMeasure();
+
+        rowCount = in.readInt();
+        indexBlockStart += Integer.BYTES;
+
+        nonNullBitmapNumber = in.readInt();
+        indexBlockStart += Integer.BYTES;
+
+        hasNullValue = in.readBoolean();
+        indexBlockStart++;
+
+        if (hasNullValue) {
+            nullValueOffset = in.readInt();
+            nullBitmapLength = in.readInt();
+            indexBlockStart += 2 * Integer.BYTES;
+        }
+
+        bitmapOffsets = new LinkedHashMap<>();
+
+        int bitmapBlockNumber = in.readInt();
+        indexBlockStart += Integer.BYTES;
+
+        indexBlocks = new ArrayList<>(bitmapBlockNumber);
+        for (int i = 0; i < bitmapBlockNumber; i++) {
+            Object key = valueReader.get();
+            int offset = in.readInt();
+            indexBlocks.add(
+                    new BitmapIndexBlock(dataType, options, key, offset, seekableInputStream));
+            indexBlockStart += measure.apply(key) + Integer.BYTES;
+        }
+
+        // bitmap body offset
+        int bitmapBodyOffset = in.readInt();
+        indexBlockStart += Integer.BYTES;
+
+        bodyStart = indexBlockStart + bitmapBodyOffset;
+    }
+
+    /** Split of all bitmap entries. */
+    class BitmapIndexBlock {
+
+        Object key;
+        int offset;
+        int serializedBytes = Integer.BYTES;
+        List<Entry> entryList;
+        Function<Object, Integer> keyBytesMapper;
+        DataType dataType;
+        SeekableInputStream seekableInputStream;
+        Options options;
+
+        void tryDeserialize() {
+            if (entryList == null) {
```

Comment:
Maybe you can take a look to https://cwiki.apache.org/confluence/display/PAIMON/PIP-25%3A+Introduce+a+key-value+file+format+for+paimon+primary+key+table

Here just something like LevelDB sst block, we can store length to avoid deserialize all entries.